src/cloud-sql-instance.ts

```diff
@@ -107,51 +115,118 @@ export class CloudSQLInstance {
     // requests to Cloud SQL Admin APIs.
-    this.ongoingRefreshPromise = this.throttle
-      ? this.throttle(this._refresh).call(this)
-      : this._refresh();
-
-    // awaits for the ongoing promise to resolve, since the refresh is
-    // completed once the promise is resolved, we just free up the reference
-    // to the promise at this point, ensuring any new call to `forceRefresh`
-    // is able to trigger a new refresh
-    await this.ongoingRefreshPromise;
-    this.ongoingRefreshPromise = undefined;
+    this.ongoingRefreshPromise = (
+      this.throttle && this.scheduledRefreshID
+        ? this.throttle(this.refreshValues).call(this)
+        : this.refreshValues()
+    )
+      // These needs to be part of the chain of promise referenced in
+      // ongoingRefreshPromise in order to avoid race conditions
+      .then((nextValues: RefreshableValues) => {
+        // in case the id at the moment of starting this refresh cycle has
+        // changed, that means that it has been canceled
+        if (currentRefreshId !== this.scheduledRefreshID) {
+          return nextValues;
+        }
+
+        // In case the refreshValues method succeeded
+        // then we go ahead and update values
+        this.updateValues(nextValues);
 
-    // Initializing the rate limiter at the end of the function so that the
-    // first refresh cycle is never rate-limited, ensuring there are 2 calls
-    // allowed prior to waiting a throttle interval.
+        this.scheduleRefresh();
+
+        // This is the end of the successful refresh chain, so now
+        // we release the reference to the ongoingRefreshPromise
+        this.ongoingRefreshPromise = undefined;
+
+        return nextValues;
+      })
+      .catch((err: unknown) => {
+        // In case there's already an active connection we won't throw
+        // refresh errors to the final user, scheduling a new refresh instead.
+        if (this.activeConnection) {
+          if (currentRefreshId === this.scheduledRefreshID) {
+            this.scheduleRefresh();
+          }
+        } else {
+          throw err as Error;
+        }
+
+        // This refresh cycle has failed, releases ref to ongoingRefreshPromise
+        this.ongoingRefreshPromise = undefined;
+      });
+
+    // The rate limiter needs to be initialized _after_ assigning a ref
+    // to ongoingRefreshPromise in order to avoid race conditions with
+    // the forceRefresh check that ensures a refresh cycle is not ongoing
     await this.initializeRateLimiter();
+
+    await this.ongoingRefreshPromise;
   }
 
-  private async _refresh(): Promise<void> {
+  // The refreshValues method will perform all the necessary async steps
+  // in order to get a new set of values for an instance that can then be
+  // used to create new connections to a Cloud SQL instance. It throws in
+  // case any of the internal steps fails.
+  private async refreshValues(): Promise<RefreshableValues> {
```

Comment:
should this be called `performRefresh` for consistency? https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/blob/main/internal/cloudsql/refresh.go#L309